docs/rfc/0696-p2p-node.adoc

```diff
@@ -0,0 +1,302 @@
+"`OPTIONAL`" in this document are to be interpreted as described in <<RFC2119>>
+and <<RFC8174>> when, and only when, they appear in all capitals, as shown here.
+
+CBOR <<RFC8949>> datatype definitions are given using the notation devised in
+CDDL <<RFC8610>>. By convention, `struct`-like datatypes are encoded as CBOR
+maps, where the map key is the zero-based numeric index of the field in
+declaration order.
+
+== Protocols
+
+The main purpose of the node is to operate the core protocol of link currently
+implemented in `librad::net::peer`, which currently consists of holding onto the
+futures du jour and dropping them at opportune moments. Furthermore all
+information in form of synchrounous calls and events obtainable internally via
+the sufficiently space-engineered tincans SHALL be made available.
+
+Besides the core protocol other functionality will be encapsulated in protocols
+which will be driven by a stream of well-defined inputs. The well-defined
+outputs of these protocols are used for effect management.
+
+=== Subroutines
+
+This section expands on the auxilliary routines driven by the node on top of the
+core protocol. Where each SHOULD allow to be disabled through configuration. A
+subroutine acts as reactor driving the protocols by feeding inputs - usually
+results of effects - into them and informed by the outputs schedules new
+effects.
+
+==== Public API
+
+For other processes to be able to talk to the node, it SHALL keep an API
+subroutine running, which offers introspection into its state. See the dedicated
+section for the surface the node will expose.
+
+==== Announcements
+
+Until superseded by <<pr-653, graftier ways>>, currently the main means of
+dissemination is through gossiping new interesting `Have`s. This is achieved by
+maintainig a persisted view on the list of identities and their current refs.
+During periodic runs a delta is created which serves as the basis of the
+announcements. In turn that list is made known to actively connected peers via
+the core protocols gossip.
+
+The delta mechanism works by recording a snapshot of the state of the monorepo,
+specifically the refs and on the next iteration diffing against the current
+state. Relevant refs from that diff can be used to announce or perform other
+periodic tasks.
+
+==== Replication Requests
+
+The core protocol implicitly tracks projects after replication, and explicitly
+with the `track` operation for remotes in the context of a `Urn`. To enable the
+same approach as for announcements where offline changes to the storage can be
+picked up at any point the node is running and online, this subroutine will
+maintain a persisted view and periodically (potentially provoked by a request
+over the <<API>>) build the delta for new tracked `Urn`s and remotes in their
+context.
+
+Note: A blocking issue for that is
+https://github.com/radicle-dev/radicle-link/issues/141[#141]
+
+==== Tracking
```

Comment:
Will this API be dynamic or will it only be specified at start up?

Reply:
So far I thought of it as purely static configuration at start-up, yet use-cases in other parts of the ecosystem could benefit from a dynamic API for it. It opens up a bit can worm wrt. to persistence of such changes.

Reply:
Added a point to #722, to keep the scope tight for this RFC, is that acceptable for now or do you want it to be addressed in here?

Reply:
Works for me :+1: